crates/pessimistic-proof-utils/src/local_exit_tree.rs

```diff
@@ -0,0 +1,262 @@
+            entry = if index & 1 == 1 {
+                H::merge(&sibling, &entry)
+            } else {
+                H::merge(&entry, &sibling)
+            };
+            index >>= 1;
+            if index < self.layers[height + 1].len() {
+                self.layers[height + 1][index] = entry;
+            } else {
+                self.layers[height + 1].push(entry);
+            }
+        }
+
+        leaf_index
+    }
+
+    /// Get the hash at the given height and index, or the empty hash if the
+    /// index is out of bounds.
+    pub fn get(&self, height: usize, index: usize) -> H::Digest {
+        *self.layers[height]
+            .get(index)
+            .unwrap_or(&self.empty_hash_at_height[height])
+    }
+
+    /// Returns the root of the tree.
+    pub fn get_root(&self) -> H::Digest {
+        let get_last_layer = |i| self.get(TREE_DEPTH - 1, i);
+        H::merge(&get_last_layer(0), &get_last_layer(1))
+    }
+
+    /// Returns an inclusion proof for the leaf at the given index.
+    pub fn get_proof(&self, leaf_index: usize) -> LocalExitTreeMerkleProof<H, TREE_DEPTH> {
+        assert!(
+            leaf_index < self.layers[0].len(),
+            "Leaf index out of bounds."
+        );
+        let mut siblings = [Default::default(); TREE_DEPTH];
+        let mut index = leaf_index;
+        for height in 0..TREE_DEPTH {
+            let sibling = self.get(height, index ^ 1);
+            siblings[height] = sibling;
+            index >>= 1;
+        }
+
+        LocalExitTreeMerkleProof { siblings }
+    }
+}
+
+impl<H, const TREE_DEPTH: usize> FromIterator<H::Digest> for LocalExitTreeData<H, TREE_DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Default + Serialize + DeserializeOwned,
+{
+    fn from_iter<T: IntoIterator<Item = H::Digest>>(iter: T) -> Self {
+        Self::from_leaves(iter.into_iter())
+    }
+}
+
+impl<H, const TREE_DEPTH: usize> Into<LocalExitTree<H, TREE_DEPTH>>
+    for &LocalExitTreeData<H, TREE_DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Default + Serialize + DeserializeOwned,
+{
+    fn into(self) -> LocalExitTree<H, TREE_DEPTH> {
+        let leaf_count = self.layers[0].len();
+        let mut frontier = [H::Digest::default(); TREE_DEPTH];
+        let mut index = leaf_count;
+        let mut height = 0;
+        while index != 0 {
+            if index & 1 == 1 {
+                frontier[height] = self.layers[height][index ^ 1];
+            }
+            height += 1;
+            index >>= 1;
+        }
+
+        LocalExitTree::from_parts(
+            leaf_count
+                .try_into()
+                .expect("usize expected to be at least 32 bits"),
+            frontier,
+        )
+    }
+}
+
+impl<H, const TREE_DEPTH: usize> LocalExitTreeMerkleProof<H, TREE_DEPTH>
+where
+    H: Hasher,
+    H::Digest: Eq + Copy + Default + Serialize + DeserializeOwned,
+{
+    /// Returns `true` iff the proof is valid for the given leaf, leaf index,
```

Comment:
```suggestion
    /// Returns `true` if the proof is valid for the given leaf, leaf index,
```

Reply:
That wasn't a typo, it means "if and only if" ;)

Reply:
Ok ! I thought it was a typo first, maybe we can put the full length version? I don't mind keeping it that way if it makes sense for everybody.